xarray/core/dataarray.py

```diff
@@ -580,7 +580,7 @@ def to_dataset(
             result = self._to_dataset_whole(name)
 
         if promote_attrs:
-            result.attrs = dict(self.attrs)
+            result.attrs = self.attrs.copy()
```

Comment:
Do all dict-like objects have a `.copy` method? Or should we use `copy.copy(...)`?

Reply:
`dict` does. :) `Mapping`, `MutableMapping` doesn't. 
It depends on how close we want `.attrs` to behave as `dict`, I've gone with being a straight up replacement to avoid any downstream issues. 
But we could use  `copy.copy` or `copy.deepcopy` instead.